File: codogram-backend/src/models/user.model.js
import mongoose,{Schema} from "mongoose";
import jwt from "jsonwebtoken" //jwt is bearer token , jiske pass token voh hi sahi h
import bcrypt from "bcrypt"

const userSchema = new Schema(
    {
        username:{
            type:String,
            required:true,
            unique:true,
            lowercase:true,
            trim:true,
            index:true//to make any field searchable
        },
        email:{
            type:String,
            required:true,
            unique:true,
            lowercase:true,
            trim:true,
        },
        fullname:{
            type:String,
            required:true,
            trim:true,
            index:true
        },
        avatar:{
            type:String,//cloudinary url
            required:true,
        },
        coverImage:{
            type:String,//cloudinary url
        },
        bio:{
            type:String,
        },
        skills:{
            type:[String],
            default:[]
        },
        snap_score:{
            type:Number,
            default:0
        },
        password:{
            type:String,
            required:[true,"Password is required"],
        },

        refreshToken:{
            type:String,
            default:"",
        },
    },{timestamps:true}// Automatically adds createdAt and updatedAt timestamps
)

userSchema.pre("save",async function (next){
    if(!this.isModified("password"))return next();// password change na hua h toh kyu hi ese encrypt kre

    this.password=await bcrypt.hash(this.password, 10)
    next()
})//it is a hook or middleware

userSchema.methods.isPasswordCorrect = async function
(password){
    return await bcrypt.compare(password,this.password)
}

userSchema.methods.generateAccessToken =function(){
    return jwt.sign
    (
    {
        _id:this._id,
        email:this.email,
        username:this.username,
        fullname:this.fullname
    },
    process.env.ACCESS_TOKEN_SECRET,
    {
        expiresIn:process.env.ACCESS_TOKEN_EXPIRY
    }
    )
}
userSchema.methods.generateRefreshToken =function(){
    return jwt.sign({
        _id:this._id,
    },
    process.env.REFRESH_TOKEN_SECRET,
    {
        expiresIn:process.env.REFRESH_TOKEN_EXPIRY
    }
)
}

export const User = mongoose.model("User",userSchema)

File: codogram_app_frontend/lib/model/user_model.dart
class UserModel {
  int? statusCode;
  Data? data;
  String? message;
  bool? success;

  UserModel({this.statusCode, this.data, this.message, this.success});

  UserModel.fromJson(Map<String, dynamic> json) {
    statusCode = json['statusCode'];
    data = json['data'] != null ? new Data.fromJson(json['data']) : null;
    message = json['message'];
    success = json['success'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['statusCode'] = this.statusCode;
    if (this.data != null) {
      data['data'] = this.data!.toJson();
    }
    data['message'] = this.message;
    data['success'] = this.success;
    return data;
  }
}

class Data {
  User? user;
  String? accessToken;
  String? refreshToken;

  Data({this.user, this.accessToken, this.refreshToken});

  Data.fromJson(Map<String, dynamic> json) {
    user = json['user'] != null ? new User.fromJson(json['user']) : null;
    accessToken = json['accessToken'];
    refreshToken = json['refreshToken'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    if (this.user != null) {
      data['user'] = this.user!.toJson();
    }
    data['accessToken'] = this.accessToken;
    data['refreshToken'] = this.refreshToken;
    return data;
  }
}

class User {
  String? sId;
  String? username;
  String? email;
  String? fullname;
  String? avatar;
  String? coverImage;
  String? bio;
  List<String>? skills;
  dynamic snapScore;
  String? createdAt;
  String? updatedAt;
  dynamic iV;

  User(
      {this.sId,
        this.username,
        this.email,
        this.fullname,
        this.avatar,
        this.coverImage,
        this.bio,
        this.skills,
        this.snapScore,
        this.createdAt,
        this.updatedAt,
        this.iV});

  User.fromJson(Map<String, dynamic> json) {
    sId = json['_id'];
    username = json['username'];
    email = json['email'];
    fullname = json['fullname'];
    avatar = json['avatar'];
    coverImage = json['coverImage'];
    bio = json['bio'];
    skills = json['skills'].cast<String>();
    snapScore = json['snap_score'];
    createdAt = json['createdAt'];
    updatedAt = json['updatedAt'];
    iV = json['__v'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['_id'] = this.sId;
    data['username'] = this.username;
    data['email'] = this.email;
    data['fullname'] = this.fullname;
    data['avatar'] = this.avatar;
    data['coverImage'] = this.coverImage;
    data['bio'] = this.bio;
    data['skills'] = this.skills;
    data['snap_score'] = this.snapScore;
    data['createdAt'] = this.createdAt;
    data['updatedAt'] = this.updatedAt;
    data['__v'] = this.iV;
    return data;
  }
}
